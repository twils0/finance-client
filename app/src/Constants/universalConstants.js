import stripeConfig from '../../../stripe.config.json';
import dbConfig from '../../../db.config.json';
import iexConfig from '../../../iex.config.json';

export const baseURLs = {
  STRIPE: stripeConfig.url,
  DB: dbConfig.url,
  IEX: iexConfig.urls.api,
  IEX_DEV: iexConfig.urls.dev,
  IEX_TERMS: iexConfig.urls.terms,
};

export const URLs = {
  CUSTOMERS: '/customers',
  USERS: '/users',
  EMAILS: '/emails',
  SECURITIES: '/securities',
  SECURITIES_SEARCH: '/securities/search',
  IEX_COMPANY: tickerCusip => `/stock/${tickerCusip}/company`,
  IEX_QUOTE: tickerCusip => `/stock/${tickerCusip}/quote`,
};

export const axiosConfig = {
  STRIPE: {
    baseURL: baseURLs.STRIPE,
    timeout: 30000,
  },
  DB: {
    baseURL: baseURLs.DB,
    timeout: 30000,
  },
  IEX: {
    baseURL: baseURLs.IEX,
    timeout: 30000,
  },
};

export const pathNames = {
  DEFAULT: '*',
  LOGIN: '/login',
  SIGN_UP: '/signup',
  CODE: '/code',
  CODE_VERIFY_EMAIL: '/code/:verificationId?',
  TERMS: '/terms',
  INTERNAL: '/internal',
  WATCHLIST: '/internal/watchlist',
  WATCHLIST_SECURITY_ID: '/internal/watchlist/:securityId?',
  ACCOUNT: '/internal/account',
};

export const requestStatusTypes = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
};
