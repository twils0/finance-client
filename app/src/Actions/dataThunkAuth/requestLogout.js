import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { formNames, buttonNames, buttonTexts } from '../../Constants/uiConstantsAccount';
import { setAuthStatus, resetAuthState } from '../dataActionsAuth';
import { resetAWSState } from '../dataActionsAWS';
import { resetAccountState } from '../dataActionsAccount';
import { resetWatchlistState } from '../dataActionsWatchlist';
import { setCurrentForm, setButtonText, setButtonVisible } from '../uiActionsAccount';

const requestLogout = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.auth.status[statusNames.LOGOUT].status !== requestStatusTypes.LOADING) {
    dispatch(
      setAuthStatus({
        id: statusNames.LOGOUT,
        status: requestStatusTypes.LOADING,
      }),
    );

    if (state.ui.internal.account.forms.current !== formNames.PROFILE) {
      dispatch(setCurrentForm({ current: formNames.PROFILE }));
    }

    if (state.ui.internal.account.buttons[buttonNames.FIRST].text !== buttonTexts.EDIT) {
      dispatch(setButtonText({ id: buttonNames.FIRST, text: buttonTexts.EDIT }));
    }

    if (state.ui.internal.account.buttons[buttonNames.SECOND].visible) {
      dispatch(setButtonVisible({ id: buttonNames.SECOND, visible: false }));
    }

    try {
      await Auth.signOut();
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(
        setAuthStatus({
          id: statusNames.LOGOUT,
          status: requestStatusTypes.ERROR,
        }),
      );

      raven.captureException(error, {
        logger: 'requestLogout',
      });

      return Promise.reject(error);
    }

    window.sessionStorage.removeSecurity('sessionTime');

    const intervalId = window.sessionStorage.getSecurity('interval');

    if (intervalId) {
      window.clearInterval(intervalId);

      window.sessionStorage.removeSecurity('interval');
    }

    dispatch(resetAWSState());
    dispatch(resetAccountState());
    dispatch(resetWatchlistState());
    dispatch(resetAuthState());

    raven.setUserContext();
  }

  return null;
};

export default requestLogout;
