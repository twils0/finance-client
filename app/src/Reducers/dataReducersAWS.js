import { combineReducers } from 'redux';

import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes } from '../Constants/dataConstantsAWS';

export const initialState = {
  status: requestStatusTypes.IDLE,
  user: null,
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_STATUS:
      return action.payload.status;
    case actionTypes.RESET_AWS_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const user = (state = initialState.user, action) => {
  switch (action.type) {
    case actionTypes.SET_AWS_USER:
      return action.payload.user;
    case actionTypes.RESET_AWS_STATE:
      return initialState.user;
    default:
      return state;
  }
};

const dataReducersAWS = combineReducers({
  status,
  user,
});

export default dataReducersAWS;
