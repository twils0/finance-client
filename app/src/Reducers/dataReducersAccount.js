import { combineReducers } from 'redux';

import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes, statusNames, fieldNames } from '../Constants/dataConstantsAccount';

export const initialState = {
  status: {
    [statusNames.AWS_FIELDS]: {
      id: statusNames.AWS_FIELDS,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.STRIPE_FIELDS]: {
      id: statusNames.STRIPE_FIELDS,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.UPDATE_STRIPE_FIELDS_TOKEN]: {
      id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.UPDATE_STRIPE_FIELDS_REQUEST]: {
      id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.UPDATE_DB_FIELDS]: {
      id: statusNames.UPDATE_DB_FIELDS,
      status: requestStatusTypes.IDLE,
    },
    [statusNames.UPDATE_AWS_FIELDS]: {
      id: statusNames.UPDATE_AWS_FIELDS,
      status: requestStatusTypes.IDLE,
    },
  },
  fields: {
    list: [...Object.values(fieldNames)],
    [fieldNames.NAME]: {
      id: fieldNames.NAME,
      value: '',
    },
    [fieldNames.EMAIL]: {
      id: fieldNames.EMAIL,
      value: '',
    },
    [fieldNames.EMAIL_ADDITIONAL]: {
      id: fieldNames.EMAIL_ADDITIONAL,
      value: '',
    },
    [fieldNames.PHONE]: {
      id: fieldNames.PHONE,
      value: '',
    },
    [fieldNames.NAME_ON_CARD]: {
      id: fieldNames.NAME_ON_CARD,
      value: '',
    },
    [fieldNames.PROMO_CODE]: {
      id: fieldNames.PROMO_CODE,
      value: '',
    },
    [fieldNames.PROMO_CODE_VALID]: {
      id: fieldNames.PROMO_CODE_VALID,
      value: true,
    },
    [fieldNames.CUSTOMER]: {
      id: fieldNames.CUSTOMER,
      value: '',
    },
  },
};

export const status = (state = initialState.status, action) => {
  switch (action.type) {
    case actionTypes.SET_ACCOUNT_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.RESET_ACCOUNT_STATE:
      return initialState.status;
    default:
      return state;
  }
};

export const fields = (state = initialState.fields, action) => {
  switch (action.type) {
    case actionTypes.SET_FIELD:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    case actionTypes.SET_FIELDS:
      return {
        ...state,
        ...action.payload,
      };
    case actionTypes.RESET_ACCOUNT_STATE:
      return initialState.fields;
    default:
      return state;
  }
};

const dataReducersAccount = combineReducers({
  status,
  fields,
});

export default dataReducersAccount;
