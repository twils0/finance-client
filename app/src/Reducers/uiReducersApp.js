import { combineReducers } from 'redux';

import stripeConfig from '../../../stripe.config.json';
import { requestStatusTypes } from '../Constants/universalConstants';
import { actionTypes, imageNames } from '../Constants/uiConstantsApp';
import {
  formNames as formNamesSignUp,
  inputNames as inputNamesSignUp,
} from '../Constants/uiConstantsExternal';
import {
  formNames as formNamesAccount,
  inputNames as inputNamesAccount,
} from '../Constants/uiConstantsAccount';

import exampleHeader from '../Images/exampleHeader.png';

export const initialState = {
  fonts: {
    status: requestStatusTypes.IDLE,
  },
  images: {
    list: [...Object.values(imageNames)],
    [imageNames.EXAMPLE_HEADER]: {
      id: imageNames.EXAMPLE_HEADER,
      src: exampleHeader,
      status: requestStatusTypes.IDLE,
    },
  },
  stripe: {
    apiKey: stripeConfig.apiKey,
    status: requestStatusTypes.IDLE,
    elements: {
      [inputNamesSignUp[formNamesSignUp.CARD].STRIPE]: {
        id: inputNamesSignUp[formNamesSignUp.CARD].STRIPE,
        loaded: false,
      },
      [inputNamesAccount[formNamesAccount.BILLING].STRIPE]: {
        id: inputNamesAccount[formNamesAccount.BILLING].STRIPE,
        loaded: false,
      },
    },
    stripeObject: null,
  },
};

export const fonts = (state = initialState.fonts, action) => {
  switch (action.type) {
    case actionTypes.SET_FONTS_STATUS:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

export const images = (state = initialState.images, action) => {
  switch (action.type) {
    case actionTypes.SET_IMAGE_STATUS:
      return {
        ...state,
        [action.payload.id]: {
          ...state[action.payload.id],
          ...action.payload,
        },
      };
    default:
      return state;
  }
};

export const stripe = (state = initialState.stripe, action) => {
  switch (action.type) {
    case actionTypes.SET_STRIPE_STATUS:
      return {
        ...state,
        ...action.payload,
      };
    case actionTypes.SET_STRIPE_ELEMENT_LOADED:
      return {
        ...state,
        elements: {
          ...state.elements,
          [action.payload.id]: {
            ...state.elements[action.payload.id],
            ...action.payload,
          },
        },
      };
    case actionTypes.SET_STRIPE_OBJECT:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

const uiReducersApp = combineReducers({
  fonts,
  images,
  stripe,
});

export default uiReducersApp;
