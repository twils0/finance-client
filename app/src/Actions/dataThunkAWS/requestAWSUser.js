import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestLogout from '../dataThunkAuth/requestLogout';

import { demo } from '../../../../mode.config.json';

const requestAWSUser = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.aws.status !== requestStatusTypes.LOADING) {
    dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

    let user = null;

    try {
      if (!demo) {
        user = await Auth.currentUserPoolUser();
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));

      if (
        (typeof error === 'string'
          && (error === 'not authenticated' || error === 'No current user in userPool'))
        || (typeof error === 'object' && error && error.code === 'NotAuthorizedException')
      ) {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'requestAWSUser',
        });
      }

      return Promise.reject(error);
    }

    if (!demo) {
      dispatch(setAWSUser({ user }));
    }
    dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));

    return { user };
  }

  return null;
};

export default requestAWSUser;
