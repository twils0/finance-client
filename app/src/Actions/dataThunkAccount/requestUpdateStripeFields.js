import axios from 'axios';
import raven from 'raven-js';

import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAccount';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import requestLogout from '../dataThunkAuth/requestLogout';
import { setAccountStatus } from '../dataActionsAccount';

import handleErrorCatch from '../../handleErrorCatch';

import { demo } from '../../../../mode.config.json';

const requestUpdateStripeFields = (payload) => {
  if (
    (!Object.prototype.hasOwnProperty.call(payload, 'element')
      || !Object.prototype.hasOwnProperty.call(payload, 'name'))
    && !Object.prototype.hasOwnProperty.call(payload, 'email')
    && !Object.prototype.hasOwnProperty.call(payload, 'promoCode')
  ) {
    throw new Error(
      `Please enter a value for either the 'name', 'email', or 'promoCode' key - ${JSON.stringify(
        payload,
      )}`,
    );
  }

  const plan = 'basic_plan';

  return async (dispatch, getState) => {
    const state = getState();

    if (
      state.data.account.status[statusNames.UPDATE_STRIPE_FIELDS_TOKEN].status
        !== requestStatusTypes.LOADING
      && state.data.account.status[statusNames.UPDATE_STRIPE_FIELDS_REQUEST].status
        !== requestStatusTypes.LOADING
    ) {
      dispatch(
        setAccountStatus({
          id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
          status: requestStatusTypes.LOADING,
        }),
      );

      let { user } = state.data.aws;

      if (!user) {
        ({ user } = await dispatch(requestAWSUser()));
      }

      let tokenId = null;

      if (!demo) {
        if (payload.element && payload.name) {
          const stripe = state.ui.app.stripe.stripeObject;

          try {
            const { token, error } = await stripe.createToken(payload.element, {
              name: payload.name,
            });

            if (error) {
              throw error;
            }

            tokenId = token.id;
          } catch (errorCatch) {
            const error = handleErrorCatch(errorCatch);

            dispatch(
              setAccountStatus({
                id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
                status: requestStatusTypes.ERROR,
              }),
            );

            raven.captureException(error, {
              logger: 'requestUpdateStripeFields',
            });

            return Promise.reject(error);
          }
        }
      }

      dispatch(
        setAccountStatus({
          id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
          status: requestStatusTypes.SUCCESS,
        }),
      );

      dispatch(
        setAccountStatus({
          id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
          status: requestStatusTypes.LOADING,
        }),
      );

      try {
        if (!demo) {
          const idToken = user.signInUserSession.idToken.jwtToken;
          const accessToken = user.signInUserSession.accessToken.jwtToken;

          await axios.put(
            URLs.CUSTOMERS,
            {
              token: tokenId,
              email: payload.email,
              promoCode: payload.promoCode,
              plan,
            },
            {
              headers: {
                Authorization: idToken,
              },
              params: {
                accessToken,
              },
              ...axiosConfig.STRIPE,
            },
          );
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);

        dispatch(
          setAccountStatus({
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.ERROR,
          }),
        );

        if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
          dispatch(requestLogout());

          return null;
        }

        if (
          typeof error === 'object'
          && error
          && (error.raw
            || error.code === 'coupon_invalid'
            || error.code === 'coupon_expired'
            || error.code === 'UsernameExistsException')
        ) {
          return Promise.reject(error);
        }

        if (!demo) {
          raven.captureException(error, {
            logger: 'requestUpdateStripeFields',
          });
        }

        return Promise.reject(error);
      }

      dispatch(
        setAccountStatus({
          id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
          status: requestStatusTypes.SUCCESS,
        }),
      );
    }

    return null;
  };
};

export default requestUpdateStripeFields;
