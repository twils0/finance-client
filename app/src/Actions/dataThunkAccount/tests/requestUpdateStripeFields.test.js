import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';

import { initialState as initialStateAWS } from '../../../Reducers/dataReducersAWS';
import { initialState as initialStateAccount } from '../../../Reducers/dataReducersAccount';
import { initialState as initialStateUIApp } from '../../../Reducers/uiReducersApp';
import { requestStatusTypes, URLs, axiosConfig } from '../../../Constants/universalConstants';
import {
  actionTypes as actionTypesAccount,
  statusNames,
} from '../../../Constants/dataConstantsAccount';
import requestLogout from '../../dataThunkAuth/requestLogout';
import requestAWSUser from '../../dataThunkAWS/requestAWSUser';
import requestUpdateStripeFields from '../requestUpdateStripeFields';

jest.mock('axios', () => ({
  put: jest.fn(),
}));
jest.mock('../../dataThunkAuth/requestLogout', () => jest.fn());
jest.mock('../../dataThunkAWS/requestAWSUser', () => jest.fn());

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const userObject = {
  user: {
    signInUserSession: {
      idToken: { jwtToken: 'testIdToken' },
      accessToken: { jwtToken: 'testAccessToken' },
    },
  },
};
const element = { test: 'testElement' };
const name = 'testName';
const stripePayloadNoCard = {
  email: 'testEmail',
  promoCode: 'testPromo',
};
const stripePayload = {
  element,
  name,
  ...stripePayloadNoCard,
};
const tokenId = 'testTokenId';
const plan = 'Basic_450';
const putPayloadNoCard = {
  token: null,
  ...stripePayloadNoCard,
  plan,
};
const putPayload = {
  token: tokenId,
  ...stripePayloadNoCard,
  plan,
};
const idToken = userObject.user.signInUserSession.idToken.jwtToken;
const accessToken = userObject.user.signInUserSession.accessToken.jwtToken;
const mockAxiosConfig = {
  headers: {
    Authorization: idToken,
  },
  params: {
    accessToken,
  },
  ...axiosConfig.STRIPE,
};

describe('dataThunkAccount', () => {
  describe('requestUpdateStripeFields', () => {
    afterEach(() => {
      requestAWSUser.mockReset();
      axios.put.mockReset();
      requestLogout.mockReset();
    });

    it("fails and throws an error when missing 'element', 'name', 'email', and/or 'promoCode' keys in payload", async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      stateBeforeAWS.user = userObject.user;
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: stateBeforeAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      axios.put.mockReturnValue(Promise.resolve());

      const emptyPayload = {};

      try {
        await store.dispatch(requestUpdateStripeFields(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(requestAWSUser).not.toBeCalled();
      expect(axios.put).not.toBeCalled();
      expect(requestLogout).not.toBeCalled();
    });

    it("calls requestAWSUser when missing 'user' object in state", async () => {
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: initialStateAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      requestAWSUser.mockReturnValue(() => Promise.resolve(userObject));
      axios.put.mockReturnValue(Promise.resolve());

      await store.dispatch(requestUpdateStripeFields(stripePayload));

      expect(requestAWSUser).toBeCalled();
      expect(stateBeforeUIApp.stripe.stripeObject.createToken).toBeCalledWith(element, { name });
      expect(axios.put).toBeCalledWith(URLs.CUSTOMERS, putPayload, mockAxiosConfig);
      expect(requestLogout).not.toBeCalled();
    });

    it('fails and returns promise reject when requestAWSUser throws an error', async () => {
      const store = mockStore({
        data: { aws: initialStateAWS, account: initialStateAccount },
        ui: { app: initialStateUIApp },
      });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      requestAWSUser.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestUpdateStripeFields(stripePayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).toBeCalled();
      expect(axios.put).not.toBeCalled();
      expect(requestLogout).not.toBeCalled();
    });

    it("calls requestLogout when axios throws 'NotAuthorizedException' error", async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      stateBeforeAWS.user = userObject.user;
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: stateBeforeAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      const error = { code: 'NotAuthorizedException', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      axios.put.mockReturnValue(Promise.reject(error));
      requestLogout.mockReturnValue(() => null);

      try {
        await store.dispatch(requestUpdateStripeFields(stripePayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(null);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(axios.put).toBeCalledWith(URLs.CUSTOMERS, putPayload, mockAxiosConfig);
      expect(requestLogout).toBeCalled();
    });

    it('fails and returns promise reject when stripe.createToken throws an error', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = {
        token: { id: tokenId },
        error: { code: 'testCode', message: 'testMessage' },
      };
      const createToken = jest.fn();
      stateBeforeAWS.user = userObject.user;
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: stateBeforeAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      const expectedActions = [
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      axios.put.mockReturnValue(Promise.resolve());

      try {
        await store.dispatch(requestUpdateStripeFields(stripePayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(tokenResult.error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(stateBeforeUIApp.stripe.stripeObject.createToken).toBeCalledWith(element, { name });
      expect(axios.put).not.toBeCalled();
      expect(requestLogout).not.toBeCalled();
    });

    it('fails and returns promise reject when axios throws an error', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      stateBeforeAWS.user = userObject.user;
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: stateBeforeAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      axios.put.mockReturnValue(Promise.reject(error));

      try {
        await store.dispatch(requestUpdateStripeFields(stripePayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(requestAWSUser).not.toBeCalled();
      expect(createToken).toBeCalledWith(element, { name });
      expect(axios.put).toBeCalledWith(URLs.CUSTOMERS, putPayload, mockAxiosConfig);
      expect(requestLogout).not.toBeCalled();
    });

    it('skips stripe.createToken when element and name key missing from payload', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      stateBeforeAWS.user = userObject.user;
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: stateBeforeAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      const expectedActions = [
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      axios.put.mockReturnValue(Promise.resolve());

      const result = await store.dispatch(requestUpdateStripeFields(stripePayloadNoCard));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestAWSUser).not.toBeCalled();
      expect(createToken).not.toBeCalled();
      expect(axios.put).toBeCalledWith(URLs.CUSTOMERS, putPayloadNoCard, mockAxiosConfig);
      expect(requestLogout).not.toBeCalled();
    });

    it('creates the correct actions with the correct payloads', async () => {
      const stateBeforeAWS = JSON.parse(JSON.stringify(initialStateAWS));
      const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
      const tokenResult = { token: { id: tokenId } };
      const createToken = jest.fn();
      stateBeforeAWS.user = userObject.user;
      createToken.mockReturnValue(Promise.resolve(tokenResult));
      stateBeforeUIApp.stripe.stripeObject = { createToken };

      const store = mockStore({
        data: { aws: stateBeforeAWS, account: initialStateAccount },
        ui: { app: stateBeforeUIApp },
      });

      const expectedActions = [
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_TOKEN,
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAccount.SET_ACCOUNT_STATUS,
          payload: {
            id: statusNames.UPDATE_STRIPE_FIELDS_REQUEST,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      axios.put.mockReturnValue(Promise.resolve());

      const result = await store.dispatch(requestUpdateStripeFields(stripePayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
      expect(requestAWSUser).not.toBeCalled();
      expect(createToken).toBeCalledWith(element, { name });
      expect(axios.put).toBeCalledWith(URLs.CUSTOMERS, putPayload, mockAxiosConfig);
      expect(requestLogout).not.toBeCalled();
    });
  });
});
