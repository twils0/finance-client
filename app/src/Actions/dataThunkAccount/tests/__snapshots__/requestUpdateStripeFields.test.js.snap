// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataThunkAccount requestUpdateStripeFields fails and throws an error when missing 'element', 'name', 'email', and/or 'promoCode' keys in payload 1`] = `[Error: Please enter a value for either the 'name', 'email', or 'promoCode' key - {}]`;
